File: python/17298_오큰수.py
n = int(input())
values = list(map(int, input().split()))
answer = [-1] * n
stack = []

stack.append(0)
for i in range(1, n):
    while stack and values[stack[-1]] < values[i]:
        answer[stack.pop()] = values[i]
    stack.append(i)

print(*answer, end='')

File: previous/Swift/BarkingDog/lecture 05 스택/17298_오큰수.swift
//
//  17298_오큰수.swift
//  Algorithm For Final Swift
//
//  Created by Kim Yewon on 2023/06/07.
//

import Foundation

let N = Int(readLine()!)!
var input = readLine()!.split { $0 == " " }.map { Int(String($0))! }
var stack = [Int]()

for i in 0 ..< N {
    while !stack.isEmpty && input[stack.last!] < input[i] {
        input[stack.removeLast()] = input[i]
    }
    stack.append(i)
}

for i in stack {
    input[i] = -1
}

print(input.map { String($0) }.joined(separator: " "))
